### ISSUE TITLE
[Level Zero] hipKernelLaunchIsNonBlocking gets stuck

### ISSUE THREAD

**Issue by pjaaskel** [CHIP-SPV/chip-spv#134]
It finishes with OpenCL. Some sort of mutex issue at deinit?

```
CHIP_BE=level0 gdb --args /home/pjaaskel/src/chip-spv/build/samples/hipKernelLaunchIsNonBlocking/hipKernelLaunchIsNonBlocking
GNU gdb (Ubuntu 12.0.90-0ubuntu1) 12.0.90
...
info: running on device Intel(R) Iris(R) Xe Graphics [0x9a49]
info: copy Host2Device
CHIP error [TID 4574] [1662633912.175776953] : hipErrorNotReady (Event Not Ready) in /home/pjaaskel/src/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:379:updateFinishStatus

CHIP error [TID 4574] [1662633912.175984574] : Caught Error: hipErrorNotReady
CHIP error [TID 4574] [1662633912.176015899] : hipErrorNotReady (Event Not Ready) in /home/pjaaskel/src/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:379:updateFinishStatus

CHIP error [TID 4574] [1662633912.176031293] : Caught Error: hipErrorNotReady
Launching kernel
Kernel launched successfully
CHIP error [TID 4574] [1662633912.177256979] : hipErrorNotReady (Event Not Ready) in /home/pjaaskel/src/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:379:updateFinishStatus

CHIP error [TID 4574] [1662633912.177287781] : Caught Error: hipErrorNotReady
CHIP error [TID 4574] [1662633912.177305132] : hipErrorNotReady (Event Not Ready) in /home/pjaaskel/src/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:379:updateFinishStatus

CHIP error [TID 4574] [1662633912.177318504] : Caught Error: hipErrorNotReady
Kernel time: 0s
PASSED!
[Thread 0x7ffff55fe640 (LWP 4579) exited]


^C
Thread 1 "hipKernelLaunch" received signal SIGINT, Interrupt.
__futex_abstimed_wait_common64 (private=128, cancel=true, abstime=0x0, op=265, expected=4578, futex_word=0x7ffff5dff910) at ./nptl/futex-internal.c:57
57	./nptl/futex-internal.c: No such file or directory.
(gdb) bt
#0  __futex_abstimed_wait_common64 (private=128, cancel=true, abstime=0x0, op=265, expected=4578, futex_word=0x7ffff5dff910) at ./nptl/futex-internal.c:57
#1  __futex_abstimed_wait_common (cancel=true, private=128, abstime=0x0, clockid=0, expected=4578, futex_word=0x7ffff5dff910) at ./nptl/futex-internal.c:87
#2  __GI___futex_abstimed_wait_cancelable64 (futex_word=futex_word@entry=0x7ffff5dff910, expected=4578, clockid=clockid@entry=0, abstime=abstime@entry=0x0, private=private@entry=128)
    at ./nptl/futex-internal.c:139
#3  0x00007ffff755b6a4 in __pthread_clockjoin_ex (threadid=140737318483520, thread_return=0x0, clockid=0, abstime=0x0, block=<optimized out>) at ./nptl/pthread_join_common.c:105
#4  0x00007ffff7f2b4fd in CHIPEventMonitor::join (this=0x5555559ac6a0) at /home/pjaaskel/src/chip-spv/src/CHIPBackend.hh:316
#5  0x00007ffff7f1eadc in CHIPBackendLevel0::uninitialize (this=0x555555571350) at /home/pjaaskel/src/chip-spv/src/backend/Level0/CHIPBackendLevel0.cc:1336
#6  0x00007ffff7e544f4 in CHIPUninitializeCallOnce () at /home/pjaaskel/src/chip-spv/src/CHIPDriver.cc:152
#7  0x00007ffff7e56f72 in std::__invoke_impl<void, void (*)()> (__f=@0x7fffffffdcf8: 0x7ffff7e544e0 <CHIPUninitializeCallOnce()>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
#8  0x00007ffff7e56f55 in std::__invoke<void (*)()> (__fn=@0x7fffffffdcf8: 0x7ffff7e544e0 <CHIPUninitializeCallOnce()>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
#9  0x00007ffff7e56f38 in std::call_once<void (*)()>(std::once_flag&, void (*&&)())::{lambda()#1}::operator()() const (this=0x7fffffffdcc8)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/mutex:776
#10 0x00007ffff7e56f14 in std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (*)()>(std::once_flag&, void (*&&)())::{lambda()#1}>(void (*&)())::{lambda()#1}::operator()() const (
    this=0x7ffff79f9760) at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/mutex:712
#11 0x00007ffff7e56ee9 in std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<void (*)()>(std::once_flag&, void (*&&)())::{lambda()#1}>(void (*&)())::{lambda()#1}::__invoke() ()
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/mutex:712
#12 0x00007ffff755ef68 in __pthread_once_slow (once_control=0x7ffff7fbad58 <Uninitialized>, init_routine=0x7ffff78aedc0 <__once_proxy>) at ./nptl/pthread_once.c:116
#13 0x00007ffff7e54bcb in __gthread_once (__once=0x7ffff7fbad58 <Uninitialized>, __func=0x7ffff78aedc0 <__once_proxy>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/x86_64-linux-gnu/c++/11/bits/gthr-default.h:700
#14 0x00007ffff7e54d15 in std::call_once<void (*)()> (__once=..., __f=@0x7fffffffdcf8: 0x7ffff7e544e0 <CHIPUninitializeCallOnce()>)
    at /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/mutex:783
#15 0x00007ffff7e54523 in CHIPUninitialize () at /home/pjaaskel/src/chip-spv/src/CHIPDriver.cc:156
#16 0x00007ffff7ebb2a7 in __hipUnregisterFatBinary (Data=0x5555559992d0) at /home/pjaaskel/src/chip-spv/src/CHIPBindings.cc:3200
#17 0x0000555555556d1f in __hip_module_dtor ()
#18 0x00007ffff750a495 in __run_exit_handlers (status=0, listp=0x7ffff76de838 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at ./stdlib/exit.c:113
#19 0x00007ffff750a610 in __GI_exit (status=<optimized out>) at ./stdlib/exit.c:143
#20 0x00007ffff74eed97 in __libc_start_call_main (main=main@entry=0x5555555563b0 <main()>, argc=argc@entry=1, argv=argv@entry=0x7fffffffdf08) at ../sysdeps/nptl/libc_start_call_main.h:74
#21 0x00007ffff74eee40 in __libc_start_main_impl (main=0x5555555563b0 <main()>, argc=1, argv=0x7fffffffdf08, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7fffffffdef8) at ../csu/libc-start.c:392
#22 0x0000555555556265 in _start ()

```

**Comment by pvelesko** [comment]
Can you please provide a backtrack on where it gets stuck along with the system configuration? 

I've tested this sample on 3 different systems without any issues so far.

**Comment by pvelesko** [comment]
I can't reproduce this issue:

Arcticus
```
ctest -R hipKernelLaunchIsNonBlocking
Test project /home/pvelesko/space/CHIP-SPV/build
    Start 372: hipKernelLaunchIsNonBlocking
1/1 Test #372: hipKernelLaunchIsNonBlocking .....   Passed   10.54 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
internal    =  10.54 sec*proc (1 test)

Total Test time (real) =  11.77 sec
```

Iris
```
ctest -R hipKernelLaunchIsNonBlocking
Test project /home/pvelesko/space/CHIP-SPV/build
    Start 372: hipKernelLaunchIsNonBlocking
1/1 Test #372: hipKernelLaunchIsNonBlocking .....   Passed   33.02 sec

100% tests passed, 0 tests failed out of 1

Label Time Summary:
internal    =  33.02 sec*proc (1 test)

Total Test time (real) =  33.54 sec
```

Which runtime are you using? @pjaaskel  
It's getting stuck on event collection, I believe. It would be good to know which event it is and why it's not firing.

**Comment by pvelesko** [comment]
@pjaaskel 

Did you run this test using `ctest -j`? I get some tests that result in a hang but only when run in parallel. If that's the case we should rename this ticket.

**Comment by pjaaskel** [comment]
Still gets stuck with level0, also when executed separately. Which BE you think is currently more stable? Perhaps we should switch to OpenCL as the default if it is? At least seems to be the case for this laptop.

**Comment by pvelesko** [comment]
OpenCL is more stable currently. Sure, we can switch to that as default which I thought it already was.

Which runtime are you using on your laptop?

**Comment by pvelesko** [comment]
You could also try this with immediate command lists.

**Comment by pvelesko** [comment]
is this still an issue?

**Comment by pjaaskel** [comment]
I cannot test it at the moment since my LZ drivers are broken, it doesn't run any of the tests. Will postpone checking it for 0.9.1 or 1.0.
